aptos-move/framework/aptos-framework/sources/fungible_asset.move

```diff
@@ -166,259 +224,333 @@ module aptos_framework::fungible_asset {
-        };
-        deposit(wallet, fa);
-        if (!ungated_transfer_allowed) {
-            set_ungated_transfer(ref, wallet, false);
-        };
+        deposit_internal(to, fa);
     }
 
-    /// Transfer `ammount` of  fungible metadata with `TransferRef` even ungated transfer is disabled.
+    /// Transfer `amount` of  fungible metadata with `TransferRef` even ungated transfer is disabled.
     public fun transfer_with_ref(
         transfer_ref: &TransferRef,
-        from: &Object<FungibleAssetWallet>,
-        to: &Object<FungibleAssetWallet>,
+        from: address,
+        to: address,
         amount: u64,
-    ) acquires FungibleAssetWallet {
+    ) acquires FungibleAssetWallet, FungibleAssetWalletEvents {
         let fa = withdraw_with_ref(transfer_ref, from, amount);
         deposit_with_ref(transfer_ref, to, fa);
     }
 
-    /// Get the underlying metadata object from `MintRef`.
-    public fun mint_ref_metadata(ref: &MintRef): Object<FungibleAssetMetadata> {
-        ref.metadata
+    /// Extract a given amount from the given fungible asset and return a new one.
+    public fun extract(fungible_asset: &mut FungibleAsset, amount: u64): FungibleAsset {
+        assert!(fungible_asset.amount >= amount, error::invalid_argument(EINSUFFICIENT_BALANCE));
+        fungible_asset.amount = fungible_asset.amount - amount;
+        FungibleAsset {
+            metadata: fungible_asset.metadata,
+            amount,
+        }
     }
 
-    /// Get the underlying metadata object from `TransferRef`.
-    public fun transfer_ref_metadata(ref: &TransferRef): Object<FungibleAssetMetadata> {
-        ref.metadata
+    /// Extract everything from the given fungible asset and return a new one.
+    public fun extract_all(fungible_asset: &mut FungibleAsset): FungibleAsset {
+        let amount = fungible_asset.amount;
+        fungible_asset.amount = 0;
+        FungibleAsset {
+            metadata: fungible_asset.metadata,
+            amount,
+        }
     }
 
-    /// Get the underlying metadata object from `BurnRef`.
-    public fun burn_ref_metadata(ref: &BurnRef): Object<FungibleAssetMetadata> {
-        ref.metadata
+    /// "Merges" the two given fungible assets. The coin passed in as `dst_fungible_asset` will have a value equal
+    /// to the sum of the two (`dst_fungible_asste` and `source_fungible_asset`).
+    public fun merge(dst_fungible_asset: &mut FungibleAsset, source_fungible_asset: FungibleAsset) {
+        let FungibleAsset { metadata: _, amount } = source_fungible_asset;
+        dst_fungible_asset.amount = dst_fungible_asset.amount + amount;
     }
 
-    /// Assert the owner of `metadata`.
-    inline fun assert_owner<T: key>(owner: &signer, metadata: &Object<T>) {
-        assert!(object::is_owner(*metadata, signer::address_of(owner)), error::permission_denied(ENOT_OWNER));
+    /// Create a new fungible asset with a value of 0.
+    public fun zero<T: key>(metadata: Object<T>): FungibleAsset {
```

Comment:
I see similar things in current coin.move. But I don't understand the potential use case. Can you share an example in comment?

Reply:
In the past, this is really useful because Coin is storable. zero() would allow people to initialize their module's key resources with empty state. It's less useful now that Coin is no longer storable, but it can make the following programming pattern easier:

let all_payments = fungible_asset::zero<T>();
collect_payment_1(&mut all_payments, ...);
collect_payment_2(&mut all_payments, ...);
collect_payment_3(&mut all_payments, ...);

What do you think?

Reply:
Then I would prefer a `merge` function in my old PR that takes a vector of fa and merge into 1 using `vector::fold`.